src/shogun/structure/MAPInference.cpp

```diff
@@ -38,6 +39,9 @@ CMAPInference::CMAPInference(CFactorGraph* fg, EMAPInferType inference_method)
 		case GRAPH_CUT:
```

Comment:
Yes, I totally agree with you. 

I am also considering to re-factor energy table and use `SGNDArray` to store energies and messages.  

Table factor operation class could also be implemented in the future because the current operation can directly done by `SGNDArray` which is not straightforward and hard to understand. Such ideas are just inspired from `DRWIN`.

What do you think?